models/settings_helpers.go

```diff
@@ -49,29 +48,6 @@ func GetSettings(q reform.DBTX) (*Settings, error) {
 func SaveSettings(q reform.DBTX, s *Settings) error {
 	s.fillDefaults()
 
-	for _, pair := range []struct {
-		dur  time.Duration
-		name string
-	}{
-		{dur: s.MetricsResolutions.HR, name: "hr"},
-		{dur: s.MetricsResolutions.MR, name: "mr"},
-		{dur: s.MetricsResolutions.LR, name: "lr"},
-	} {
-		if pair.dur < time.Second {
-			return status.Error(codes.InvalidArgument, pair.name+": minimal resolution is 1s")
-		}
-		if pair.dur.Truncate(time.Second) != pair.dur {
-			return status.Error(codes.InvalidArgument, pair.name+": should be a natural number of seconds")
-		}
-	}
-
-	if s.DataRetention < 24*time.Hour {
-		return status.Error(codes.InvalidArgument, "data_retention: minimal resolution is 24h")
-	}
-	if s.DataRetention.Truncate(24*time.Hour) != s.DataRetention {
-		return status.Error(codes.InvalidArgument, "data_retention: should be a natural number of days")
-	}
```

Comment:
One step forward, one step back…

We have two ways to set some settings: via environment variables, and via API. **Both** should validate values. **Both** should return proper error messages, referencing environment variable names, or API field names, respectively. But actual validation (for example, that data retention interval is a natural number of days) should happen in **one** place. That code should not be copy&pasted in two places.

Now environment variables are checked, but API tests fail: https://pmm.cd.percona.com/blue/organizations/jenkins/pmm2-api-tests/detail/pmm2-api-tests/2193/tests

@BupycHuk, please check that during the next review. Please also check that FB was updated.

Reply:
Don't forget about this endpoint https://github.com/percona/pmm-managed/pull/341/files#diff-6f891b9f28cfb10751fe182afa544a9cR487